### PR TITLE
Add PHP streaming test generation

### PR DESCRIPTION
**Pull request by michaelbausor**
Staging PR: https://github.com/googleapis/api-client-staging/pull/341

### CODE REVIEW

**Comment by garrettjonesgoogle** [comment]
@shinfan could you take a look first?

**Comment by michaelbausor** [comment]
Updated to match changes in https://github.com/googleapis/api-client-staging/pull/341, PTAL

**Comment by michaelbausor** [comment]
Cleaned up and merged master branch, PTAL

File: src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java [hunk 1]
```diff
@@ -96,18 +97,52 @@ public TestCaseView createTestCaseView(
...
     }
 
+    InitCodeContext responseInitCodeContext =
+        createResponseInitCodeContext(methodContext, initCodeContext.symbolTable());
+    MockGrpcResponseView mockGrpcResponseView =
+        createMockResponseView(methodContext, responseInitCodeContext);
+
+    GrpcStreamingView grpcStreamingView = GrpcStreamingView.newBuilder().build();
+    if (methodConfig.isGrpcStreaming()) {
+
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java]: nit: remove empty line

File: src/main/java/com/google/api/codegen/transformer/php/PhpGapicSurfaceTestTransformer.java [hunk 2]
```diff
@@ -212,11 +213,19 @@ private static ModelTypeTable createTypeTable(String packageName) {
     for (Method method : context.getSupportedMethods()) {
       GapicMethodContext methodContext = context.asRequestMethodContext(method);
 
-      if (methodContext.getMethodConfig().isGrpcStreaming()) {
-        // TODO(shinfan): Remove this check once grpc streaming is supported by test
+      if (methodContext.getMethodConfig().getGrpcStreamingType()
+          == GrpcStreamingType.ClientStreaming) {
+        //TODO: Add unit test generation for ClientStreaming methods
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/php/PhpGapicSurfaceTestTransformer.java]: space before TODO

File: src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java [hunk 3]
```diff
@@ -233,6 +282,13 @@ private InitCodeContext createResponseInitCodeContext(
         additionalSubTrees.add(InitCodeNode.createSingletonList(subResponseFieldName));
       }
     }
+    if (context.getMethodConfig().isGrpcStreaming()) {
+      GrpcStreamingConfig grpcConfig = context.getMethodConfig().getGrpcStreaming();
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java]: s/grpcConfig/streamingConfig/

File: src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java [hunk 4]
```diff
@@ -96,18 +97,52 @@ public TestCaseView createTestCaseView(
...
+    if (methodConfig.isGrpcStreaming()) {
+
+      String resourceTypeName = null;
+      String resourcesFieldGetterName = null;
+      if (methodConfig.getGrpcStreaming().hasResourceField()) {
+        Field resourcesField = methodConfig.getGrpcStreaming().getResourcesField();
+        resourceTypeName =
+            methodContext.getTypeTable().getAndSaveNicknameForElementType(resourcesField.getType());
+        resourcesFieldGetterName =
+            methodContext
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java]: You can reuse the namer variable at L60

File: src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java [hunk 5]
```diff
@@ -96,18 +97,52 @@ public TestCaseView createTestCaseView(
           !ServiceMessages.s_isEmptyType(methodConfig.getLongRunningConfig().getReturnType());
     }
 
+    InitCodeContext responseInitCodeContext =
+        createResponseInitCodeContext(methodContext, initCodeContext.symbolTable());
+    MockGrpcResponseView mockGrpcResponseView =
+        createMockResponseView(methodContext, responseInitCodeContext);
+
+    GrpcStreamingView grpcStreamingView = GrpcStreamingView.newBuilder().build();
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java]: You can default the grpc streaming view to null if it is not a grpc streaming method:
GrpcStreamingView grpcStreamingView;

if (methodConfig.isGrpcStreaming()) {
  grpcStreamingView = GrpcStreamingView.newBuilder().build();
}

File: src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java (hunk 1 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java]: Done

File: src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java (hunk 4 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java]: Done

File: src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java (hunk 3 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java]: Done

File: src/main/java/com/google/api/codegen/transformer/php/PhpGapicSurfaceTestTransformer.java (hunk 2 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/php/PhpGapicSurfaceTestTransformer.java]: Done

File: src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java (hunk 5 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java]: Defaulted to null

File: src/main/java/com/google/api/codegen/viewmodel/testing/GrpcStreamingView.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
+/* Copyright 2016 Google Inc
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/viewmodel/testing/GrpcStreamingView.java]: 2017

**Comment by shinfan** [comment]
@michaelbausor I can not see the updates, did you push the change?

**Comment by michaelbausor** [comment]
I can't see them either, which is super weird. They are showing up in my fork, but not in the PR... (https://github.com/michaelbausor/toolkit/commits/streaming-test-gen)

**Comment by michaelbausor** [comment]
OK well closing and re-opening seems to have fixed the issue, no idea how that happened...

File: src/main/java/com/google/api/codegen/viewmodel/testing/TestCaseView.java [hunk 7]
```diff
@@ -33,6 +34,10 @@
 
   public abstract List<ClientTestAssertView> asserts();
 
+  public boolean hasAsserts() {
```
Comment (shinfan) [inline src/main/java/com/google/api/codegen/viewmodel/testing/TestCaseView.java]: Is this used anywhere?

File: src/main/java/com/google/api/codegen/viewmodel/testing/GrpcStreamingView.java (hunk 6 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/viewmodel/testing/GrpcStreamingView.java]: Done

File: src/main/java/com/google/api/codegen/viewmodel/testing/TestCaseView.java (hunk 7 above)
Comment (michaelbausor) [inline src/main/java/com/google/api/codegen/viewmodel/testing/TestCaseView.java]: No, this is unused, well spotted :-)

This was used in the snippets, but is not needed. Removed.

**Comment by michaelbausor** [comment]
This is some seriously weird stuff, need to close and reopen the PR each time I push for it to pick up changes...

**Pull request merged**